src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -2082,13 +2082,21 @@ private BoundExpression BindNonMethod(SimpleNameSyntax node, Symbol symbol, Bind
                         {
                             Error(diagnostics, ErrorCode.ERR_InvalidPrimaryConstructorParameterReference, node, parameter);
                         }
-                        else if (parameter.IsExtensionParameter() &&
-                                (InParameterDefaultValue || InAttributeArgument ||
-                                 this.ContainingMember() is not { Kind: not SymbolKind.NamedType, IsStatic: false } || // We are not in an instance member
-                                 (object)this.ContainingMember().ContainingSymbol != parameter.ContainingSymbol) &&
-                                !IsInsideNameof)
+                        else if (parameter.IsExtensionParameter() && !IsInsideNameof)
                         {
-                            Error(diagnostics, ErrorCode.ERR_InvalidExtensionParameterReference, node, parameter);
+                            var member = this.ContainingMember();
+                            var isInSameExtension = (object)member.ContainingSymbol == parameter.ContainingSymbol;
+                            var isStaticMember = member is { Kind: not SymbolKind.NamedType, IsStatic: true };
+                            var isInstanceMember = member is { Kind: not SymbolKind.NamedType, IsStatic: false };
```

Comment:
>var isInstanceMember = member is { Kind: not SymbolKind.NamedType, IsStatic: false }; [](http://example.com/codeflow?start=28&length=85)

This doesn't match any part of the logic in the original condition, therefore, the code that is using it below cannot  be considered as a pure refactoring. I prefer not going through an exercise of proving that the code is going to result in the same outcome as the previous code for all scenarios possible today or in the future, modulo the error number. Please restructure the logic so that it would be "in original terms".

Reply:
FWIW, the original code here was extremely difficult to follow. At the very least, I want to apply demorgan's law to the pattern to make it more readable; it is composed of multiple nots and implicit ands that would be much more readable as a simple `or` pattern.